### PR TITLE
Added parser for Nback and GNGo

### PR DESCRIPTION
**Pull request by ThomasHMAC**

### CODE REVIEW

File: bin/dm_parse_GNGo.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import pandas as pd
+import subprocess
+import glob
+import sys
+
+arguments       = docopt(__doc__)
+inputfile        = arguments['<inputfile>']
+resultsfile      = arguments['--output']
+DEBUG           = arguments['--debug']
+DRYRUN          = arguments['--dry-run']
```
Comment (jerdra) [inline bin/dm_parse_GNGo.py:40]: can you run a linter/code formatter for this? Dawn and I have been using "yapf" to re-format our code!

File: bin/dm_parse_GNGo.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+if DEBUG : print(arguments)
```
Comment (jerdra) [inline bin/dm_parse_GNGo.py:42]: can you indent here? running a re-formatter could work as well

File: bin/dm_parse_GNGo.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+## loop over the inputfiles
+
+if DEBUG: print("\n file: {}".format(inputfile))
```
Comment (jerdra) [inline bin/dm_parse_GNGo.py:46]: same as above re:indenting

File: bin/dm_parse_GNGo.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+import codecs
+import sys
+import pdb
+from docopt import docopt
+import numpy as np
+import os
+import pandas as pd
+import subprocess
+import glob
+import sys
```
Comment (jerdra) [inline bin/dm_parse_GNGo.py:34]: Might be a good idea to use "import logging" to print to stdout. Would be worth checking other datman/bin scripts and seeing how they log!

File: bin/dm_parse_GNGo.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        if (("ImageDisplay1.RT:" in line) and ("ImageDisplay1.RTTime:" not in line)):
+            ReactionTime = line.split(':')[1].replace(" ",'')
+        if "ImageDisplay1.RESP:" in line:
+            SubjectResponse = line.split(':')[1].replace(" ",'')
+        if "ImageDisplay1.CRESP:" in line:
+            CorrectResponse = line.split(':')[1].replace(" ",'')
+        if "image:" in line:
+            StimFile = line.split(':')[1].replace(" ",'')
+    #Skip if list type
+    if (TrialType == ''):continue
```
Comment (jerdra) [inline bin/dm_parse_GNGo.py:87]: indent here

File: bin/dm_parse_Nback.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+## loop over individual trials
+trial_count = 0
+for trial in trials:
+
+    # for each trial collect the CorrectResponse, Subject's Response and the Reation Time
+    CorrectResponse = int()
+    ReactionTime = int()
+    SubjectResponse = int()
+    TrailType = str()
+    TrailTime = int()
```
Comment (jerdra) [inline bin/dm_parse_Nback.py:68]:
```suggestion
    TrialType = str()
    TrialTime = int()
```

File: bin/dm_parse_Nback.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+    for line in trial.split('\n\t'):
+        if ("Running:" in line) :
+            TrailType = line.split(':')[1].replace(" ",'')
```
Comment (jerdra) [inline bin/dm_parse_Nback.py:72]:
```suggestion
            TrialType = line.split(':')[1].replace(" ",'')
```

File: bin/dm_parse_Nback.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        if ("letterdisp.OnsetTime:" in line) :
+            TrailTime = line.split(':')[1].replace(" ",'')
```
Comment (jerdra) [inline bin/dm_parse_Nback.py:74]:
```suggestion
            TrialTime = line.split(':')[1].replace(" ",'')
```

File: bin/dm_parse_Nback.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        if ("resp:" in line) and ("letterdisp.RESP" not in line) and ("letterdisp.CRESP" not in line):
+            CorrectResponse = line.split(':')[1].replace(" ",'')
+        if (("letterdisp.RT" in line) and ("letterdisp.RTTime" not in line)):
+            ReactionTime = line.split(':')[1].replace(" ",'')
+        if "letterdisp.RESP" in line:
+            SubjectResponse = line.split(':')[1].replace(" ",'')
+
+    #Skip if list type
+    if (TrailType == 'List1') or (TrailType == ''):continue
```
Comment (jerdra) [inline bin/dm_parse_Nback.py:83]:
```suggestion
    if (TrialType == 'List1') or (TrialType == ''):continue
```

Indent would be needed here as well

File: bin/dm_parse_Nback.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+import glob
+import sys
+
+arguments       = docopt(__doc__)
+inputfile        = arguments['<inputfile>']
+resultsfile      = arguments['--output']
+DEBUG           = arguments['--debug']
+DRYRUN          = arguments['--dry-run']
+
+if DEBUG : print(arguments)
```
Comment (jerdra) [inline bin/dm_parse_Nback.py:37]: indent here

File: bin/dm_parse_Nback.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+            ReactionTime = line.split(':')[1].replace(" ",'')
+        if "letterdisp.RESP" in line:
+            SubjectResponse = line.split(':')[1].replace(" ",'')
+
+    #Skip if list type
+    if (TrailType == 'List1') or (TrailType == ''):continue
+    trial_count += 1
+
+    #Form trial dictionary
+    trial_dict = {'onset' : TrailTime, 'duration' : 0,
```
Comment (jerdra) [inline bin/dm_parse_Nback.py:87]:
```suggestion
    trial_dict = {'onset' : TrialTime, 'duration' : 0,
```

File: bin/dm_parse_Nback.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+                 'trial_type' : TrailType, 'correct_response' : corr[str(CorrectResponse)],
```
Comment (jerdra) [inline bin/dm_parse_Nback.py:88]:
```suggestion
                 'trial_type' : TrialType, 'correct_response' : corr[str(CorrectResponse)],
```

**Review by jerdra — CHANGES_REQUESTED** [review]
Thanks for setting this up Thomas! Just a few stylistic comments I've added. Would recommend that you try a linter/code formatter like yapf to adjust the code style to match what we've set out as code style standards.

Other thing to consider maybe is whether we can refactor these scripts into a single script/library of scripts to handle task parsing since there're a ton of overlaps b/w the two scripts. Obv having something working for the moment is important - but something to think about for sure!

**Comment by ThomasHMAC** [comment]
> Thanks for setting this up Thomas! Just a few stylistic comments I've added. Would recommend that you try a linter/code formatter like yapf to adjust the code style to match what we've set out as code style standards.
> 
> Other thing to consider maybe is whether we can refactor these scripts into a single script/library of scripts to handle task parsing since there're a ton of overlaps b/w the two scripts. Obv having something working for the moment is important - but something to think about for sure!

Thanks for the suggestion Jer, and yes there are tons of overlapping between two scripts. I will try to make a single script/ library of scripts later

File: bin/dm_parse_GNGo.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+    print(arguments)
+
+# Loop over the inputfiles
+if DEBUG:
+    print("\n file: {}".format(inputfile))
+
+try:
+    with open(inputfile, "r", encoding="utf-16") as myfile:
+        filetext = myfile.read()
+except myfile.DoesNotExist:
```
Comment (DESm1th) [inline bin/dm_parse_GNGo.py:44]: This should be updated to 

```
except FileNotFoundError:
```

**Review by DESm1th — CHANGES_REQUESTED** [review]
Looks really good, thank you Thomas! Like Jer I just have a few stylistic python-nitpicks, but I'm comfortable approving things as they are now once the exception is fixed. 

1. In python it's good to use snake-case for multi-word variable and function names so for instance "inputfile" would be "input_file" instead. This is a PEP8 practice, but most linters can't check for it.

2. It's a good idea to put code inside a main function / other functions (and then call the main function at the bottom). Every instruction that's not inside of a function gets executed when the script is imported, which can make it difficult or impossible to write tests for code. If it's in a function, though, it doesn't execute until the function is called (with a standard 'guard' at the bottom of the script like below, the main function will always be called if the script is being run directly on the command line and not if it's imported somewhere else)
```
def main():
    # your code here

if __name__ == "__main__":
    main()
```

File: bin/dm_parse_GNGo.py [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+message_df["response_time"] = message_df["response_time"].apply(
+    lambda x: x / 1000
+)
+
+# Load in the gngo onset time csv file
+try:
+    df_onset = pd.read_csv("gng_event_timing.csv")
+except df_onset.DoesNotExist:
+    print("ERROR opening onset time file")
+    quit()
```
Comment (jerdra) [inline bin/dm_parse_GNGo.py:123]: exception should be fixed here as well FileNotFoundError

**Review by jerdra — CHANGES_REQUESTED** [review]

**Review by DESm1th — APPROVED** [review]
Thanks Thomas! Good work :D

**Review by jerdra — APPROVED** [review]

**Pull request merged**